### PR TITLE
feat(monitoring): ship Grafana dashboards as pod-scoped ConfigMaps

### PR DESCRIPTION
**Pull request by christopherfickess**
#### Summary

**Proof-of-concept (draft) for SRE review — the "monitoring *with* Grafana dashboards" variant.**

Builds on #466 (rules + ServiceMonitor) and adds the Grafana dashboard layer, so the two branches differ **only** by this feature — letting SRE weigh adopting dashboards separately from rules/ServiceMonitor.

- Adds `spec.monitoring.grafanaDashboard` (enable + sidecar discovery label/value) and the reconcile path.
- Ships **all nine Mattermost dashboards** from `mattermost-performance-assets`, transformed for Kubernetes by `hack/transform_dashboards.py`:
  - **Datasource normalization** — collapse the per-export `${DS_*}` inputs onto one templated `datasource` variable and drop `__inputs`/`__requires`, so the Grafana sidecar provisions them without an import prompt.
  - **instance → pod** — the four server dashboards now select a stable `pod` (scoped by a new `namespace` variable) instead of the churning `instance` IP:port (123 `instance=~"$server"` filters rewritten to `pod=~"$server"`).
- **One ConfigMap per dashboard** (largest ~115 KB) to stay under the 1 MB ConfigMap limit and match the sidecar's one-file-per-ConfigMap convention.
- **Cleanup/prune** — each dashboard ConfigMap carries a stable marker label; disabling the flag deletes them all, and reconcile prunes ones no longer in the embedded set (rename/removal), independent of the configurable discovery label.

**Validation:** unit tests; an `envtest` reconcile test asserting the ConfigMaps are created and deleted; and a live `kind` + kube-prometheus-stack run confirming the stock Grafana sidecar imports and provisions **all nine** dashboards (`/api/search`).

Dashboard-only delta vs the base branch:

```
git diff feat/operator-monitoring-poc..feat/operator-monitoring-poc-grafana
```

> Note: the Calls dashboard's rtcd/offloader/node-exporter (`:9100`) panels intentionally remain instance-based — they depend on those separate exporters. Client-app dashboards (web/mobile/desktop) need `clientMetrics` enabled to have data.

#### Ticket Link

_Internal SRE monitoring POC — no public ticket. Companion to #466._

#### Release Note

```release-note
Add `spec.monitoring.grafanaDashboard` to the Mattermost CRD: ships the Mattermost Grafana dashboards as pod-scoped ConfigMaps (one per dashboard) for pickup by the Grafana dashboard sidecar.
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by christopherfickess** [comment]
Thanks for the review. This PR builds on #466; the shared findings (service-label scoping, `MattermostLabels` map copy, license-warning scope, openapi regen, gofmt, CEL removal for K8s v1.22, server-side apply for the oversized CRD) are addressed there and included here.

**Dashboard-specific:**
- Dashboard ConfigMaps carry a stable marker label (`mattermost.com/grafana-dashboard`) independent of the configurable discovery label, and cleanup/prune lists by that marker — so only operator-owned ConfigMaps are deleted.
- Removed the now-obsolete CEL envtest (the `callsMetrics` CEL validation was dropped for CI-kind compatibility); the reconcile envtest remains.

**Deferred (POC):** the `up`-with-no-endpoints edge case and interval/label pattern validation — noted as follow-ups.

File: test/envtest/cel_validation_test.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
+package envtest_test
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply test/envtest/cel_validation_test.go]: Fixed — the CEL `x-kubernetes-validations` was removed (CI kind is K8s 1.22, which rejects it) and this envtest was deleted. rtcdServiceSelector is still enforced at runtime.

File: pkg/mattermost/monitoring.go [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+// GenerateMetricsServiceV1Beta builds a dedicated, internal (headless ClusterIP)
+// Service exposing only the Mattermost metrics port. The ServiceMonitor targets
+// THIS Service rather than the main app Service, so scraping works in every
+// service mode — including useServiceLoadBalancer, where the app Service exposes
+// only 80/443 and drops the metrics port. Metrics stay internal to the cluster;
+// they are never published through the LoadBalancer.
+func GenerateMetricsServiceV1Beta(mattermost *mmv1beta.Mattermost) *corev1.Service {
+	labels := mattermost.MattermostLabels(mattermost.Name)
+	// Distinct label so the ServiceMonitor selects only this Service.
+	labels[metricsServiceLabel] = mattermost.Name
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply pkg/mattermost/monitoring.go:87]: Fixed at the root — `MattermostLabels` now copies `spec.resourceLabels` instead of aliasing it.

File: config/rbac/role.yaml [hunk 3]
```diff
@@ -119,3 +119,19 @@ rules:
       - update
       - patch
       - delete
+  # Optional: manage ServiceMonitors when spec.monitoring.serviceMonitor is enabled.
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply config/rbac/role.yaml:122]: Fixed — added `events` create/patch to the operator ClusterRole so the recorder can emit the warning.

File: controllers/mattermost/mattermost/monitoring.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+		mattermost.Spec.Monitoring.CallsMetrics != nil &&
+		mattermost.Spec.Monitoring.CallsMetrics.Enabled
+}
+
+// anyMonitoringEnabled reports whether any monitoring capability that depends on
+// the (Enterprise-gated) /metrics endpoint is turned on.
+func anyMonitoringEnabled(mattermost *mmv1beta.Mattermost) bool {
+	return serviceMonitorEnabled(mattermost) ||
+		prometheusRuleEnabled(mattermost) ||
+		callsMetricsEnabled(mattermost)
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply controllers/mattermost/mattermost/monitoring.go]: Fixed — `callsMetrics` is excluded from the license warning; rtcd metrics are not gated by the Mattermost license.

File: pkg/mattermost/prometheusrules/mattermost.yaml [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+# NOTE: the /metrics endpoint is Enterprise-licensed — the `up`/mattermost_*
+# alerts only evaluate when a valid license is installed. Thresholds are a
+# conservative starting point and usually want per-environment tuning.
+groups:
+  - name: mattermost.availability
+    rules:
+      - alert: MattermostServerDown
+        # Metrics target unscrapable for 5m (server down, or metrics disabled /
+        # unlicensed so /metrics on 8067 is not served).
+        expr: up{namespace="__NAMESPACE__", service="__SERVICE__"} == 0
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply pkg/mattermost/prometheusrules/mattermost.yaml:27]: Documented the no-`up`-series limitation and why we avoid `absent()` in rule-only configs; pair with the crash-loop / replica alerts.

File: hack/transform_dashboards.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    templating = dashboard.setdefault("templating", {}).setdefault("list", [])
+
+    # Does this dashboard select a Mattermost server (instance-based)? If so it
+    # gets the full pod-scoping treatment.
+    server_var = next((v for v in templating if v.get("name") == "server"), None)
+    is_server_dashboard = server_var is not None
+
+    if is_server_dashboard:
+        podify_server_var(server_var)
+        dashboard = replace_instance_filter(dashboard)
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply hack/transform_dashboards.py:249]: Acknowledged (intentional for the POC). Only the four server dashboards use `instance=~\"$server\"`; the calls/client panels use different label sets (rtcd/offloader, agent/platform) and are out of scope for the instance→pod rewrite — noted in the PR description.

File: hack/transform_dashboards.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    if isinstance(node, dict):
+        return {k: replace_ds_strings(v) for k, v in node.items()}
+    if isinstance(node, list):
+        return [replace_ds_strings(v) for v in node]
+    if isinstance(node, str):
+        if node in PROM_DS_TOKENS:
+            return "${%s}" % DATASOURCE_VAR_NAME
+        if node == EXPR_DS_TOKEN:
+            return EXPR_DS_UID
+        return node
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply hack/transform_dashboards.py:66]: Acknowledged — we normalize the known `${DS_*}` inputs these exports actually use; a general datasource pass is a follow-up if new dashboards introduce other input names.

File: pkg/mattermost/monitoring.go [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+// (the combined dashboards exceed 400KB) and matches the sidecar's
+// one-file-per-ConfigMap convention. Created in the Mattermost namespace; a
+// Grafana running elsewhere reads them (Grafana -> Mattermost).
+func GenerateGrafanaDashboardConfigMapsV1Beta(mattermost *mmv1beta.Mattermost) ([]*corev1.ConfigMap, error) {
+	discoveryLabel := defaultDashboardDiscoveryLabel
+	discoveryValue := defaultDashboardDiscoveryLabelValue
+
+	if gd := mattermost.Spec.Monitoring.GrafanaDashboard; gd != nil {
+		if gd.DiscoveryLabel != "" {
+			discoveryLabel = gd.DiscoveryLabel
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply pkg/mattermost/monitoring.go:253]: Acknowledged — the marker key is namespaced under `mattermost.com/`, so a user setting `discoveryLabel` to that exact key is not a realistic config. Marker is applied after the discovery label; pruning keys off the marker.

File: Makefile [hunk 9]
```diff
@@ -122,6 +125,17 @@ unittest: ## Runs unit tests
 e2e-local:
 	./test/e2e_local.sh
 
+# Fast local validation with envtest (a real apiserver+etcd from binaries — no
+# Docker). Covers CRD CEL validation and the monitoring reconcile create/delete
+# lifecycle. Downloads the apiserver binaries on first run.
+ENVTEST_K8S_VERSION ?= 1.31.0
+test-envtest: $(SETUP_ENVTEST)
+	KUBEBUILDER_ASSETS="$$($(SETUP_ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(TOOLS_BIN_DIR)/envtest -p path)" \
+		$(GO) test -tags envtest ./test/envtest/... ./controllers/mattermost/mattermost/... -run 'CEL|Monitoring' -v
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply Makefile]: Fixed — the `test-envtest` target no longer passes the removed `./test/envtest/...` package (and dropped the obsolete `CEL` run-filter); it now runs the reconcile envtest under `./controllers/...`. Verified locally: `ok ... 6.6s`.

File: controllers/mattermost/mattermost/monitoring.go [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+	if err := r.Client.List(context.TODO(), existing,
+		client.InNamespace(mattermost.Namespace),
+		client.MatchingLabels(mattermostApp.DashboardConfigMapSelector(mattermost)),
+	); err != nil {
+		return errors.Wrap(err, "failed to list grafana dashboard config maps")
+	}
+
+	if !grafanaDashboardEnabled(mattermost) {
+		return r.deleteConfigMaps(existing.Items, reqLogger)
+	}
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply controllers/mattermost/mattermost/monitoring.go:65]: Fixed — deleteConfigMaps now filters with metav1.IsControlledBy, so an unrelated ConfigMap carrying the discovery/marker label is skipped (logged), never deleted.

File: pkg/resources/create_resources.go [hunk 11]
```diff
@@ -126,6 +128,103 @@ func (r *ResourceHelper) CreateServiceIfNotExists(owner v1.Object, service *core
...
+	err := r.client.Get(context.TODO(), key, found)
+	if err != nil {
+		if apimeta.IsNoMatchError(err) || k8sErrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrap(err, "failed to check if service monitor exists")
+	}
+
+	reqLogger.Info("Deleting service monitor", "name", key.Name)
+	if err := r.client.Delete(context.TODO(), found); err != nil && !k8sErrors.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply pkg/resources/create_resources.go]: Fixed — cleanup goes through deleteOwnedResource (IsControlledBy check); the name-only DeleteServiceMonitor/DeletePrometheusRule helpers were removed.

File: controllers/mattermost/mattermost/monitoring.go [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+// gracefully (logged skip) when the Prometheus Operator CRDs are absent.
+func (r *MattermostReconciler) checkMattermostServiceMonitor(mattermost *mmv1beta.Mattermost, reqLogger logr.Logger) error {
+	metricsSvcKey := types.NamespacedName{Name: mattermost.Name + "-metrics", Namespace: mattermost.Namespace}
+	smKey := types.NamespacedName{Name: mattermost.Name, Namespace: mattermost.Namespace}
+
+	if !serviceMonitorEnabled(mattermost) {
+		if err := r.Resources.DeleteServiceMonitor(smKey, reqLogger); err != nil {
+			return err
+		}
+		return r.Resources.DeleteService(metricsSvcKey, reqLogger)
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply controllers/mattermost/mattermost/monitoring.go]: Fixed — the metrics Service cleanup uses deleteOwnedResource and only deletes when this Mattermost is the controller owner.

File: controllers/mattermost/mattermost/monitoring.go [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+	err = r.Client.Get(context.TODO(), ruleKey, current)
+	if err != nil {
+		// CRD absent (creation was skipped) or not yet present — nothing to update.
+		if apimeta.IsNoMatchError(err) || k8sErrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrap(err, "failed to fetch current prometheus rule")
+	}
+
+	return r.Resources.Update(current, desired, reqLogger)
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply controllers/mattermost/mattermost/monitoring.go:200]: Fixed — the PrometheusRule update is now guarded by ensureOwnedForUpdate (IsControlledBy), same as the ConfigMap guard.

File: controllers/mattermost/mattermost/monitoring.go [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: desired.Name, Namespace: desired.Namespace}, current)
+	if err != nil {
+		// CRD absent (creation was skipped) or not yet present — nothing to update.
+		if apimeta.IsNoMatchError(err) || k8sErrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrap(err, "failed to fetch current service monitor")
+	}
+
+	return r.Resources.Update(current, desired, reqLogger)
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply controllers/mattermost/mattermost/monitoring.go:245]: Fixed — reconcileServiceMonitor guards the update with ensureOwnedForUpdate, protecting both the main and rtcd paths.

File: controllers/mattermost/mattermost/monitoring.go [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+
+	current := &corev1.Service{}
+	if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: desired.Name, Namespace: desired.Namespace}, current); err != nil {
+		return errors.Wrap(err, "failed to fetch current metrics service")
+	}
+
+	if err := ensureOwnedForUpdate(mattermost, current); err != nil {
+		return err
+	}
+	return r.Resources.Update(current, desired, reqLogger)
```
[the remark this answers is not in the document]
Comment (christopherfickess) [reply controllers/mattermost/mattermost/monitoring.go:222]: Fixed — `reconcileService` now calls `resources.CopyServiceEmptyAutoAssignedFields(desired, current)` before updating, preserving the API-assigned clusterIP/clusterIPs (["None"] for the headless metrics Service), same as the main Service reconcile. Applied on all three branches.
